### PR TITLE
Fix bug so that `add_type` updates forward references in new type added 

### PR DESCRIPTION
**Pull request by huong-li-nguyen**
## Description
- Fix bug on `ChildWithForwardRef.update_forward_refs` having to be called
- Add unit tests to capture new test case
- Update docs

## Screenshot
Without the fix, the unit tests will fail on this:
Screenshot 2023-10-04 at 13 04 27

## Checklist

- [x] I have not referenced individuals, products or companies in any commits, directly or indirectly
- [x] I have not added data or restricted code in any commits, directly or indirectly
- [x] I have updated the docstring of any public function/class/model changed
- [x] I have added the PR number to the change description in the changelog fragment, e.g. `Enable feature XXX ([#1](https://github.com/mckinsey/vizro/pull/1))` (if applicable)
- [x] I have added tests to cover my changes (if applicable)

## Types of changes

- [x] Docs/refactoring (non-breaking change which improves codebase)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Notice

- [x] I acknowledge and agree that, by checking this box and clicking "Submit Pull Request":

  - I submit this contribution under the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0.txt) and represent that I am entitled to do so on behalf of myself, my employer, or relevant third parties, as applicable.
  - I certify that (a) this contribution is my original creation and / or (b) to the extent it is not my original creation, I am authorised to submit this contribution on behalf of the original creator(s) or their licensees.
  - I certify that the use of this contribution as authorised by the Apache 2.0 license does not violate the intellectual property rights of anyone else.

### CODE REVIEW

File: vizro-core/changelog.d/20231004_131631_huong_li_nguyen_fix_child_forwardref.md [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+<!--
+### Deprecated
+
+- A bullet item for the Deprecated category.
+
+-->
+
+### Fixed
+
+- Fix bug on `ChildWithForwardRef.update_forward_refs()` having to be called explicitly ([#92](https://github.com/mckinsey/vizro/pull/92))
```
Comment (antonymilne) [inline vizro-core/changelog.d/20231004_131631_huong_li_nguyen_fix_child_forwardref.md:34]:
```suggestion
- Fix bug so that `add_type` updates forward references in new type added ([#92](https://github.com/mckinsey/vizro/pull/92))
```

Comment (antonymilne) [reply vizro-core/changelog.d/20231004_131631_huong_li_nguyen_fix_child_forwardref.md:34]: I haven't read or understood our scheme for writing the changelog yet, so ignore me if this isn't in keeping with our usual tone... but usually a good rule of thumb is that the things we write should make some sense to the end user. Since `ChildWithForwardRef` doesn't exist outside our test suite, it won't mean anything to user.

File: vizro-core/tests/unit/vizro/models/test_base.py [hunk 2]
```diff
@@ -19,6 +19,11 @@ class ChildZ(vm.VizroBaseModel):
     type: Literal["child_Z"] = "child_Z"
 
 
+class ChildWithForwardRef(vm.VizroBaseModel):
+    type: Literal["child_forward_ref"] = "child_forward_ref"
```
Comment (antonymilne) [inline vizro-core/tests/unit/vizro/models/test_base.py:23]:
```suggestion
    type: Literal["child_with_forward_ref"] = "child_with_forward_ref"
```

File: vizro-core/tests/unit/vizro/models/test_base.py [hunk 3]
```diff
@@ -151,6 +156,27 @@ def test_add_type_dict_instantiation(self, ParentWithForwardRef, mocker):
         assert isinstance(parent.child, ChildZ)
 
 
+class TestChildForwardRefDiscriminatedUnion:
```
Comment (antonymilne) [inline vizro-core/tests/unit/vizro/models/test_base.py:159]:
```suggestion
class TestChildWithForwardRef:
```

File: vizro-core/tests/unit/vizro/models/test_base.py [hunk 4]
```diff
@@ -151,6 +156,27 @@ def test_add_type_dict_instantiation(self, ParentWithForwardRef, mocker):
...
+        child = ChildWithForwardRef()
+        with pytest.raises(ValidationError, match="No match for discriminator 'type' and value 'child_forward_ref'"):
+            Parent(child=child)
+
+    def test_add_type_model_instantiation(self, Parent, mocker):
+        # Make it as if these are in vizro.models so that update_forward_refs call in add_type works on them.
+        mocker.patch.dict(vm.__dict__, {"ChildXForwardRef": ChildX})
+        Parent.add_type("child", ChildWithForwardRef)
+        parent = Parent(child=ChildWithForwardRef(grandchild=ChildX()))
+        assert isinstance(parent.child, ChildWithForwardRef)
```
Comment (antonymilne) [inline vizro-core/tests/unit/vizro/models/test_base.py:170]:
```suggestion
        assert isinstance(parent.child, ChildWithForwardRef) and isinstance(parent.child.grandchild, ChildX)
```

Doesn't matter much but would be slightly stricter.

File: vizro-core/tests/unit/vizro/models/test_base.py [hunk 5]
```diff
@@ -151,6 +156,27 @@ def test_add_type_dict_instantiation(self, ParentWithForwardRef, mocker):
...
+
+    def test_add_type_dict_instantiation(self, Parent, mocker):
+        # Make it as if these are in vizro.models so that update_forward_refs call in add_type works on them.
+        mocker.patch.dict(vm.__dict__, {"ChildXForwardRef": ChildX})
+        Parent.add_type("child", ChildWithForwardRef)
+        parent = Parent(child={"type": "child_forward_ref", "grandchild": {"type": "child_x"}})
+        assert isinstance(parent.child, ChildWithForwardRef)
```
Comment (antonymilne) [inline vizro-core/tests/unit/vizro/models/test_base.py:177]:
```suggestion
assert isinstance(parent.child, ChildWithForwardRef) and isinstance(parent.child.grandchild, ChildX)
```

File: vizro-core/tests/unit/vizro/models/test_base.py [hunk 6]
```diff
@@ -151,6 +156,27 @@ def test_add_type_dict_instantiation(self, ParentWithForwardRef, mocker):
...
+        mocker.patch.dict(vm.__dict__, {"ChildXForwardRef": ChildX})
+        Parent.add_type("child", ChildWithForwardRef)
+        parent = Parent(child=ChildWithForwardRef(grandchild=ChildX()))
+        assert isinstance(parent.child, ChildWithForwardRef)
+
+    def test_add_type_dict_instantiation(self, Parent, mocker):
+        # Make it as if these are in vizro.models so that update_forward_refs call in add_type works on them.
+        mocker.patch.dict(vm.__dict__, {"ChildXForwardRef": ChildX})
+        Parent.add_type("child", ChildWithForwardRef)
+        parent = Parent(child={"type": "child_forward_ref", "grandchild": {"type": "child_x"}})
```
Comment (antonymilne) [inline vizro-core/tests/unit/vizro/models/test_base.py:176]:
```suggestion
        parent = Parent(child={"type": "child_forward_ref", "grandchild": {}})
```
We shouldn't need to specify `type` here. If this change doesn't work then please tell me though because it means my understanding of something is wrong...

File: vizro-core/tests/unit/vizro/models/test_base.py (hunk 2 above)
Comment (antonymilne) [reply vizro-core/tests/unit/vizro/models/test_base.py:23]: (would need find and replace throughout the file since it's in some other places too)

**Review by antonymilne — APPROVED** [review]
Thank you for finishing this off! 💯

File: vizro-core/changelog.d/20231004_131631_huong_li_nguyen_fix_child_forwardref.md (hunk 1 above)
Comment (huong-li-nguyen) [reply vizro-core/changelog.d/20231004_131631_huong_li_nguyen_fix_child_forwardref.md:34]: Yes, that makes a lot of sense! I like the phrasing as well 👍

File: vizro-core/tests/unit/vizro/models/test_base.py (hunk 6 above)
Comment (huong-li-nguyen) [reply vizro-core/tests/unit/vizro/models/test_base.py]: Ah yes, that works!

File: vizro-core/docs/assets/tutorials/dashboard/dashboard23.png
Comment (maxschulz-COL) [inline vizro-core/docs/assets/tutorials/dashboard/dashboard23.png:1]: why this change?

**Review by maxschulz-COL — APPROVED** [review]
Lgtm! Great fix

File: vizro-core/docs/assets/tutorials/dashboard/dashboard23.png
Comment (huong-li-nguyen) [reply vizro-core/docs/assets/tutorials/dashboard/dashboard23.png:1]: The initial screenshots were created when the left-side panel wasn't automatically removed. The old screenshot outdated and shows the left-side panel without any controls or navigations, so I updated them!
